search/floydwarshall.go

```diff
@@ -5,226 +5,168 @@
+		next: make([][]int, len(nodes)*len(nodes)),
+	}
+	for i, u := range nodes {
+		paths.dist.Set(i, i, 0)
+		for _, v := range from(u) {
+			j := indexOf[v.ID()]
+			paths.set(i, j, weight(edgeTo(u, v)), j)
 		}
+	}
 
-		// This step is a tad convoluted: we have some list of intermediates.
-		// We can think of each intermediate as a path junction
-		//
-		// At this junction, we can find all the shortest paths back to i,
-		// and all the shortest paths down to j. Since this is a junction,
-		// any predecessor path that runs through this intermediate may
-		// freely choose any successor path to get to j. They'll all be
-		// of equivalent length.
-		//
-		// Thus, for each intermediate, we run through and join each predecessor
-		// path with each successor path via its junction.
-		for _, intermediate := range intermediates {
-
-			// Find predecessors
-			preds, err := allPathFinder(i, intermediate)
-			if err != nil {
-				return nil, err
-			}
-
-			// Join each predecessor with its junction
-			for a := range preds {
-				preds[a] = append(preds[a], nodes[intermediate])
-			}
-
-			// Find successors
-			succs, err := allPathFinder(intermediate, j)
-			if err != nil {
-				return nil, err
-			}
-
-			// Join each successor with its predecessor at the junction.
-			// (the copying stuff is because slices are reference types)
-			for a := range succs {
-				for b := range preds {
-					path := make([]graph.Node, len(succs[a]), len(succs[a])+len(preds[b]))
-					copy(path, succs[a])
-					path = append(path, preds[b]...)
-					toReturn = append(toReturn, path)
+	for k := range nodes {
+		for i := range nodes {
+			for j := range nodes {
+				ij := paths.dist.At(i, j)
+				joint := paths.dist.At(i, k) + paths.dist.At(k, j)
+				if ij > joint {
+					paths.set(i, j, joint, paths.at(i, k)...)
+				} else if i != k && k != j && ij-joint == 0 {
+					paths.add(i, j, paths.at(i, k)...)
 				}
 			}
-
 		}
-
-		return toReturn, nil
 	}
 
-	return func(start, goal graph.Node) ([][]graph.Node, float64, error) {
-		paths, err := allPathFinder(start.ID(), goal.ID())
-		if err != nil {
-			return nil, inf, err
+	ok = true
+	for i := range nodes {
+		if paths.dist.At(i, i) < 0 {
+			ok = false
+			break
 		}
+	}
 
-		for i := range paths {
-			// Prepend start and postpend goal, but don't repeat start/goal
-
-			if len(paths[i]) != 0 {
-				if paths[i][0].ID() != start.ID() {
-					paths[i] = append(paths[i], nil)
-					copy(paths[i][1:], paths[i][:len(paths[i])-1])
-					paths[i][0] = start
-				}
+	return paths, ok
+}
 
-				if paths[i][len(paths[i])-1].ID() != goal.ID() {
-					paths[i] = append(paths[i], goal)
-				}
-			} else {
-				paths[i] = append(paths[i], start, goal)
-			}
-		}
+// ShortestPaths is a shortest path tree created by the FloydWarshall function.
+type ShortestPaths struct {
+	nodes   []graph.Node
+	indexOf map[int]int
 
-		return paths, dist[start.ID()+goal.ID()*numNodes], nil
-	}
+	dist *mat64.Dense
+	next [][]int
```

Comment:
I would add a comment explaining what next[i] contains. It is obvious to me now, but the understanding was definitely not immedate. The algorithm then follows easily.